framework/src/interfaces/SetupInterface.C

```diff
@@ -33,6 +33,14 @@ SetupInterface::SetupInterface(const MooseObject * moose_object)
         (moose_object->parameters().getCheckedPointerParam<FEProblemBase *>("_fe_problem_base"))
             ->getCurrentExecuteOnFlag())
 {
+  // PETSc no longer calls SNESComputeFunction at the end of a linear solve, so we do this so that
+  // users who expect things to be updated at the end of a solve are satisfied
+  FEProblemBase & fe_problem =
+      *moose_object->parameters().getCheckedPointerParam<FEProblemBase *>("_fe_problem_base");
+  if (fe_problem.solverParams()._type == Moose::ST_LINEAR)
+    if (_execute_enum.contains(EXEC_LINEAR))
+      const_cast<ExecFlagEnum &>(_execute_enum) += EXEC_TIMESTEP_END;
```

Comment:
Just a question: is this safe if timestep_end has been set in the input by the users?

Reply:
I would suspect that it wouldn't matter though because I'd expect the execution flags would be queried via `MultiMooseEnum::contains`, which should work with duplicates.